lib/Echidna/Processor.hs

```diff
@@ -0,0 +1,133 @@
+filterPayable = map g . filter f
+                 where f (PayableInfo _) = True
+                       f _               = False
+                       g (PayableInfo i) = i
+                       g _               = error "fail in filterPayable"
+
+filterConstantFunction :: [SlitherInfo] -> [(T.Text, [T.Text])]
+filterConstantFunction = map g . filter f 
+                 where f (ConstantFunctionInfo _) = True
+                       f _               = False
+                       g (ConstantFunctionInfo i) = i
+                       g _               = error "fail in filterConstantFunction"
+
+
+filterGenerationGraph :: [SlitherInfo] -> [(T.Text, T.Text, [T.Text])]
+filterGenerationGraph = map g . filter f
+                 where f (GenerationGraph _) = True
+                       f _                   = False
+                       g (GenerationGraph i) = i
+                       g _                   = error "fail in filterGenerationGraph"
+
+-- Slither processing
+runSlither :: (MonadIO m, MonadThrow m) => FilePath -> [String] -> m [SlitherInfo]
+runSlither fp aa = let args = ["--ignore-compile", "--print", "echidna", "--json", "-"] ++ aa in do
+  mp  <- liftIO $ findExecutable "slither"
+  case mp of
+   Nothing -> return [] 
+   Just path -> liftIO $ do (ec, out, err) <- readCreateProcessWithExitCode (proc path $ args |> fp) {std_err = Inherit} ""
+                            case ec of
+                              ExitSuccess -> return $ procSlither out
+                              ExitFailure _ -> throwM $ ProcessorFailure "slither" err
+
+procSlither :: String -> [SlitherInfo]
+procSlither r = case (decode . BSL.pack) r of
+                 Nothing -> []
+                 Just v  -> mresult "" v
+
+-- parse result json
+mresult :: T.Text -> Value -> [SlitherInfo]
+mresult "description" (String x)  = case (decode . BSL.pack . T.unpack) x of
+                                      Nothing -> []
+                                      Just v  -> mpayable "" v ++ mcfuncs "" v ++ mggraph "" v
+
+mresult _ (Object o) = concatMap (uncurry mresult) $ M.toList o
+mresult _ (Array  a) = concatMap (mresult "") a
+mresult _  _         = []
+
+-- parse actual payable information
+mpayable :: T.Text -> Value -> [SlitherInfo]
+mpayable "payable" (Object o)  = map ( PayableInfo . second f) $ M.toList o
+                                 where f (Array xs)            = concatMap f xs
+                                       f (String "fallback()") = ["()"]
+                                       f (String s)            = [s]
+                                       f _                     = []
+mpayable _ (Object o) = concatMap (uncurry mpayable) $ M.toList o
+mpayable _ (Array  a) = concatMap (mpayable "") a
+mpayable _  _         = []
+
+-- parse actual constant functions information
+mcfuncs :: T.Text -> Value -> [SlitherInfo]
+mcfuncs "constant_functions" (Object o)  = map ( ConstantFunctionInfo . second f) $ M.toList o
```

Comment:
rogue space

Reply:
we can also probably do `ConstantFunctionInfo . second f <$> M.toList o`

Reply:
Fixed.